pypiserver/backend.py

```diff
@@ -0,0 +1,224 @@
+        that name (yet). `fh` is an open file object that can be used to read
+        the file's content. To convert the package into an actual file on disk,
+        run `as_file(filename, fh)`.
+        """
+        raise NotImplementedError
+
+    def remove_package(self, pkg: PkgFile):
+        """Remove a package from the Backend"""
+        raise NotImplementedError
+
+    def exists(self, filename) -> bool:
+        """Does a package by the given name exist?"""
+        raise NotImplementedError
+
+    def digest(self, pkg: PkgFile):
+        if self.hash_algo is None:
+            return None
+        return digest_file(pkg.fn, self.hash_algo)
+
+    def get_projects(self) -> t.Iterable[str]:
+        """Return an iterable of all (unique) projects available in the store
+        in their PEP503 normalized form. When implementing a Backend class,
+        either use this method as is, or override it with a more performant
+        version.
+        """
+        normalized_pkgnames = set()
+        for x in self.get_all_packages():
+            if x.pkgname:
+                normalized_pkgnames.add(x.pkgname_norm)
+        return normalized_pkgnames
+
+    def find_project_packages(self, project: str) -> t.Iterable[PkgFile]:
+        """Find all packages from a given project. The project may be given
+        as either the normalized or canonical name. When implementing a
+        Backend class, either use this method as is, or override it with a
+        more performant version.
+        """
+        return (
+            x
+            for x in self.get_all_packages()
+            if normalize_pkgname(project) == x.pkgname_norm
+        )
+
+    def find_version(self, name, version) -> t.Iterable[PkgFile]:
+        """Return all packages that match PkgFile.pkgname == name and
+        PkgFile.version == version` When implementing a Backend class,
+        either use this method as is, or override it with a more performant
+        version.
+        """
+        return filter(
+            lambda pkg: pkg.pkgname == name and pkg.version == version,
+            self.get_all_packages(),
+        )
+
+
+class SimpleFileBackend(Backend):
+    def __init__(self, config: Configuration, roots: t.List[PathLike]):
+        super().__init__(config)
+        self.roots = [Path(root).resolve() for root in roots]
+
+    def get_all_packages(self):
+        return itertools.chain.from_iterable(listdir(r) for r in self.roots)
+
+    def add_package(self, filename: str, fh: t.BinaryIO):
+        as_file(fh, self.roots[0].joinpath(filename))
+
+    def remove_package(self, pkg: PkgFile):
+        os.remove(pkg.fn)
+
+    def exists(self, filename):
+        # TODO: Also look in subdirectories?
+        return any(root.joinpath(filename).exists() for root in self.roots)
+
+
+class CachingFileBackend(SimpleFileBackend):
+    def __init__(
+        self, config: Configuration, roots: t.List[PathLike], cache_manager
```

Comment:
We may want to define either a base class or a [protocol](https://docs.python.org/3/library/typing.html#typing.Protocol) for the cache_manager

Reply:
Alas, no `Protocol`s in python<3.8. I see that I put the type_hint for cache_manager in the `__init__` assignment rather that the function signature, so I'm changing that. 

Did you mean to add a sort of CacheManager interface/ base implementation that different backends may consume and/or extend? This might be an idea, but I'm afraid that there is not going to be much common requirements between the different backends, since most of the logic will lie in cache invalidation, which is going to be very backend-specific

Reply:
> Alas, no Protocols in python<3.8

This is true, but I think that if we want to use it, you can get it for earlier pythons by installing [typing-extensions](https://pypi.org/project/typing-extensions/).

> Did you mean to add a sort of CacheManager interface/ base implementation that different backends may consume and/or extend

Yep, that was the idea, so that you could potentially have like a `WatchdogCachingFileBackend` and a `WhateverOtherCachingFileBackend`, so long as the mangers share the same interface. It seems like all
we need for that at the moment is `listdir()` and `digest_file()`, so it might not be too bad to specify an abstract interface